File: src/com/easy/ImplementQueueUsingStacks_232.java
/**
 * 
 */
package com.easy;

import java.util.Stack;

/**
 * @author Chris Lin
 * @version 1.0
 */
@SuppressWarnings("unchecked")

public class ImplementQueueUsingStacks_232 {
    
    // Use utility lib stack:
    // Time Complexity: NA
    // Space Complexity: NA
    // Runtime: 41 ms, faster than 98.06%
    // Memory Usage: 33.4 MB, less than 57.77% 

    /**
     * Your MyQueue object will be instantiated and called as such: MyQueue obj
     * = new MyQueue(); obj.push(x); int param_2 = obj.pop(); int param_3 =
     * obj.peek(); boolean param_4 = obj.empty();
     */

    @SuppressWarnings("rawtypes")
    class MyQueue {
        private Stack mInStack;
        private Stack mOutStack;

        /** Initialize your data structure here. */
        public MyQueue() {
            mInStack = new Stack();
            mOutStack = new Stack();
        }

        /** Push element x to the back of queue. */
        public void push(int x) {
            mInStack.push(x);
        }

        /**
         * Removes the element from in front of queue and returns that element.
         */
        public int pop() {
            in2OutStack();
            return (int) mOutStack.pop();
        }

        /** Get the front element. */
        public int peek() {
            in2OutStack();
            return (int) mOutStack.peek();

        }

        private void in2OutStack() {
            if (mOutStack.isEmpty()) {
                while (!mInStack.isEmpty()) {
                    mOutStack.push(mInStack.pop());
                }
            }
        }

        /** Returns whether the queue is empty. */
        public boolean empty() {
            return (mInStack.isEmpty() && mOutStack.isEmpty());

        }
    }
}

File: Golang_src/easy/implementQueueUsingStacks_232.go
package easy

// Use two stack:
// Time Complexity: O(1)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 100.00%
type MyQueue struct {
	in, out []int
}

/** Initialize your data structure here. */
func ConstructorMyQueue() MyQueue {
	return MyQueue{}
}

func (this *MyQueue) in2Out() {
	for len(this.in) > 0 {
		this.out = append(this.out, this.in[len(this.in)-1])
		this.in = this.in[:len(this.in)-1]
	}
}

/** Push element x to the back of queue. */
func (this *MyQueue) Push(x int) {
	this.in = append(this.in, x)
}

/** Removes the element from in front of queue and returns that element. */
func (this *MyQueue) Pop() int {
	if this.Empty() {
		return -1
	}

	if len(this.out) == 0 {
		this.in2Out()
	}

	val := this.out[len(this.out)-1]
	this.out = this.out[:len(this.out)-1]

	return val
}

/** Get the front element. */
func (this *MyQueue) Peek() int {
	if this.Empty() {
		return -1
	}

	if len(this.out) == 0 {
		this.in2Out()
	}

	return this.out[len(this.out)-1]
}

/** Returns whether the queue is empty. */
func (this *MyQueue) Empty() bool {
	return len(this.out) == 0 && len(this.in) == 0
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param_2 := obj.Pop();
 * param_3 := obj.Peek();
 * param_4 := obj.Empty();
 */
